### PR TITLE
Molecule testing, CI refacto

### PR DESCRIPTION
**Pull request by hugoShaka**
This PR adds molecule suppport:
- add molecule using docker
- add 2 senarios
- generate internal DNS container
- move fixtures to tests files
- add azure pipeline support
- move travis CI to molecule
- various idempotency fixes
- Fix postgresql not starting because of an invalid service name
- Add flake 8 lint

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,35 +21,52 @@ DO NOT APPLY IF YOU ARE NOT CONFIDENT WITH WHAT THE PLAYBOOK DOES.
...
 * Clone the repository
 * `cd ansible-mailserver`
-* *If you've installed ansible in a venv make sure it is activated here*
+* *If you want to use a virtual environment make sure it is activated here*
+
+You can use vagrant :
+
 * `vagrant up` will create 3 VMs : the mailservers (north and south) and the tester
 
+Or you can use molecule dans docker :
```
Comment (Badidzetai) [inline README.md]: Faute ici de langue
```suggestion
Or you can use molecule in docker :
```

File: roles/ispmail-certificate/tasks/main.yml [hunk 2]
```diff
@@ -6,17 +6,23 @@
...
 - name: Create a self-signed certificate
-  command: >
-    openssl req -newkey rsa:4096
-      -nodes -sha512 -x509
-      -days "{{ ispmail_certificate_days_valid }}"
-      -nodes
-      -subj "/C={{ ispmail_certificate_country }}/ST={{ ispmail_certificate_state }}/L={{ ispmail_certificate_city }}/O={{ ispmail_certificate_organisation }}/OU={{ ispmail_certificate_orgunit }}{% for domain in ispmail_certificate_domains %}/CN={{ domain }}{% endfor %}/emailAddress={{ ispmail_certificate_email }}"
-      -out "{{ ispmail_certificate_location }}"
-      -keyout "{{ ispmail_certificate_keyfile }}"
+  command: "openssl req -newkey rsa:4096 \
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply roles/ispmail-certificate/tasks/main.yml]: It's legacy code (before openssl_* modules) , I'll add the issue to refactor the role.

File: test/suite/tools.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+def insert_virtual_users(server_address: str, users: List):
+    """Add users fixtures into the database."""
+
+    do_query(server_address, ADD_USER_QUERY, users)
+
+
+def do_query(server_address: str, query: str, items: List):
+    """Opens a connection and executes a database query."""
+
+    # TODO(shaka) gather password using testinfra and ansible info
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply test/suite/tools.py]: Thanks :)

File: test/suite/test_smtp.py [hunk 4]
```diff
@@ -85,20 +102,28 @@ def test_external_alias(server):
...
 def test_invalid_sender(server):
-    assert smtp_error_code("obiwan@jedi.local", sender="vader@none", server=server) == 504
+    assert smtp_error_code("luke@jedi.local", sender="vader@none", server=server) == 504
 
 
 def test_non_existant_sender_domain(server):
     """Should be 450 4.1.8 Sender address rejected: Domain not found"""
-    assert smtp_error_code("obiwan@jedi.local", sender="vader@poudlard.local", server=server) == 450
+    assert (
+        smtp_error_code("luke@jedi.local", sender="vader@poudlard.local", server=server)
```
Comment (toadjaune) [inline test/suite/test_smtp.py]:
```suggestion
        smtp_error_code("luke@jedi.local", sender="vader@hogwarts.local", server=server)
```
You french :P

File: molecule/create.yml [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        security_opts: "{{ item.security_opts | default(omit) }}"
+        volumes: "{{ item.volumes | default(omit) }}"
+        tmpfs: "{{ item.tmpfs | default(omit) }}"
+        capabilities: "{{ item.capabilities | default(omit) }}"
+        exposed_ports: "{{ item.exposed_ports | default(omit) }}"
+        published_ports: "{{ item.published_ports | default(omit) }}"
+        ulimits: "{{ item.ulimits | default(omit) }}"
+        networks: "{{ item.networks | default(omit) }}"
+        network_mode: "{{ item.network_mode | default(omit) }}"
+        dns_servers: "{{ internal_dns_ip }}"
```
Comment (toadjaune) [inline molecule/create.yml]: To give weight to my 2 remarks regarding this excessive templating : you really have to look for this in order to find it, it's totally hidden in the middle of these useless options. (And I knew what I was looking for)

File: molecule/create.yml [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    - name: Add mail.not.local DNS entry
+      lineinfile:
+        line: "{{ item.ansible_facts.docker_container.NetworkSettings.Networks.external.Gateway }}   mail.not.local"
+        state: present
+        path: /var/dns-hosts/molecule
+        create: true
+      loop: "{{docker_jobs.results}}"
+      delegate_to: ns
+
+    - name: Add per-container DNS entry
```
Comment (toadjaune) [inline molecule/create.yml]: I wasn't convinced when you first explained it to me, but I've got to admit, this way out of the cyclic dependency between DNS server IP, and other container's IPs is pretty elegant.

However there's one thing I don't understand : how does the config get reloaded ? Does dnsmasq constantly watch its configuration file for changes ?

File: molecule/prepare.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+---
+- name: Prepare
```
Comment (toadjaune) [inline molecule/prepare.yml]: I'm not familiar enough with molecule to know how standard this is, but to me, this kind of preparation should be directly in the docker image.

File: molecule/default/molecule.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    command: /sbin/init
+    networks:
+      - name: external
+
+  - name: south
+    groups:
+      - mta
+      - mda
+      - test
+    image: minimum2scp/systemd-stretch
```
Comment (toadjaune) [inline molecule/default/molecule.yml]: I don't know these images. Is there a specific reason to choose them over official images ?

File: molecule/Dockerfile.j2 [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# Molecule managed
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply molecule/Dockerfile.j2]: You're right it's part of default molecule.
I don't exclude adding compatibility to redhat/centos ;)

File: molecule/create.yml (hunk 5 above)
Comment (hugoShaka) [reply molecule/create.yml]: If I remove all those items from default molecule `create.yml` it'll basically break the molecule abstraction. According to the molecule documentation these variables could be defined in the `molecule.yml` file. Hardcoding variable means molecule will not behave as intended and could be weird. 

But I completely see your point of readability. It should at least be highlighted with a comment or something.

File: molecule/create.yml (hunk 6 above)
Comment (hugoShaka) [reply molecule/create.yml]: Yes, dnsmasq does listen for host files and does not need reloading. This should be documented with a comment.

File: molecule/default/molecule.yml (hunk 8 above)
Comment (hugoShaka) [reply molecule/default/molecule.yml]: Good question, @HadrienPatte what do you think ? Why did you prefer those ?

Comment (HadrienPatte) [reply molecule/default/molecule.yml]: We need systemd-enabled images, and as far as I know, the official debian images are not.

File: molecule/create.yml [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      with_items: "{{ molecule_yml.platforms | selectattr('name', 'eq', 'ns') | list }}"
+      async: 7200
+      poll: 0
+
+    - name: Wait for DNS creation to complete
+      async_status:
+        jid: "{{ item.ansible_job_id }}"
+      register: dnsdocker_jobs
+      until: dnsdocker_jobs.finished
+      retries: 300
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply molecule/create.yml]: I'll try to keep as close as possible on the initial `create.yml` (https://github.com/ansible/molecule/blob/master/molecule/provisioner/ansible/playbooks/docker/create.yml)

File: molecule/prepare.yml (hunk 7 above)
Comment (hugoShaka) [reply molecule/prepare.yml]: I don't really get where you'd like to move it :sweat_smile:.
We can't put it in the create.yml as the inventory is not yet generated. Should we maintain our own systemd images ? I'm not sure it worths the pain.

File: test/mailserver-vagrant.yml [hunk 11]
```diff
@@ -13,7 +13,8 @@
         path: /etc/fstab
         backup: true
         backrefs: true
-        regexp: '^(\S+\s+{{ point }}\s+\S+\s+)(?!(?:\S*,)?{{ opt }}(?:,\S*)?\s+)(\S+)(\s+.+)$'
+        regexp: "^(\\S+\\s+{{ point }}\\s+\\S+\\s+)(?!(?:\\S*,)?{{ opt }}\
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply test/mailserver-vagrant.yml]: I just tried, but it looks like:
```regexp
'^(\S+\s+/\s+\S+\s+)(?!(?:\S*,)?acl\(?:,\S*)?\s+)(\S+)(\s+.+)$'
```
I'm not sure it really is more understandable as we don't really see the mount point and the goal of this line anymore.

File: molecule/docker-images/Dockerfile-dnsmasq [hunk 12]
```diff
@@ -0,0 +1,5 @@
+FROM andyshinn/dnsmasq:2.75
+
+RUN mkdir /var/dns-hosts
+
+ADD dnsmasq.conf /etc/dnsmasq.conf
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply molecule/docker-images/Dockerfile-dnsmasq]: Done !

File: molecule/create.yml (hunk 6 above)
Comment (hugoShaka) [reply molecule/create.yml]: This behaviour is now described in a comment

File: test/mailserver-vagrant.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (toadjaune) [reply test/mailserver-vagrant.yml]: You're right, it's just plain ugly, and even worse than the original.
So bad ansible mountpoint management module would not allow to add an option to an existing options set.

File: molecule/create.yml [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        image: "{{ item.pre_build_image | default(false) | ternary('', 'molecule_local/') }}{{ item.image }}"
+        state: started
+        recreate: false
+        log_driver: json-file
+        command: "{{ item.command | default('bash -c \"while true; do sleep 10000; done\"') }}"
+        privileged: "{{ item.privileged | default(omit) }}"
+        security_opts: "{{ item.security_opts | default(omit) }}"
+        volumes: "{{ item.volumes | default(omit) }}"
+        tmpfs: "{{ item.tmpfs | default(omit) }}"
+        capabilities: "{{ item.capabilities | default(omit) }}"
```
[the remark this answers is not in the document]
Comment (toadjaune) [reply molecule/create.yml]: Like many other points discussed here, it's molecule standard.

File: molecule/create.yml [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        exposed_ports: "{{ item.exposed_ports | default(omit) }}"
```
[the remark this answers is not in the document]
Comment (toadjaune) [reply molecule/create.yml]: Like many other points here, it's molecule standard.

**Pull request merged**